### PR TITLE
SPINEDEM-2121 Set NHS Number Allocation permissions

### PR DESCRIPTION
**Pull request by davidhamill1-nhs**
## Summary
Place permissions step in the Apigee request flows so that *only* healthcare worker can send an NHS Number Allocation request.

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: specification/personal-demographics.yaml [hunk 1]
```diff
@@ -2096,16 +2096,6 @@ components:
...
-    XRequestID:
-      name: X-Request-ID
-      required: true
-      schema:
-        type: string
-        pattern: '^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$'
-        example: 60E0B220-8136-4CA5-AE46-1D97EF59D068
-      in: header
-      description: |
-        The X-Request-ID from the request header, if supplied, mirrored back.
```
Comment (davidhamill1-nhs) [inline specification/personal-demographics.yaml:2108]: This is unused code. The specs actually use the file XRequestID.yaml to build the documentation for the X-Request-ID response header.

File: tests/functional/test_patient_access.py [hunk 2]
```diff
@@ -1,28 +1,12 @@
...
-    access_token = _nhsd_apim_auth_token_data.get("access_token", "")
-
-    headers = {
-        "X-Request-ID": str(uuid.uuid1()),
-        "X-Correlation-ID": str(uuid.uuid1()),
-        "Authorization": f'Bearer {access_token}'
-    }
-    return headers
-
-
```
Comment (davidhamill1-nhs) [inline tests/functional/test_patient_access.py:25]: Moved to conftest to be used by app-restricted tests.

File: tests/functional/features/nhs_number_allocate.feature [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		And INVALID_METHOD is at issue[0].details.coding[0].code in the response body
+		And Cannot create resource with application-restricted access token is at issue[0].details.coding[0].display in the response body
+		And Your app has insufficient permissions to use this method. Please contact support. is at issue[0].diagnostics in the response body
+
+	Scenario: A healthcare worker can allocate an NHS number
+		Given I am a healthcare worker user
+		
+		When I request an NHS number allocation
+
+		Then I get a 400 HTTP response code
```
Comment (davidhamill1-nhs) [inline tests/functional/features/nhs_number_allocate.feature:32]:
```suggestion
		Then I do not get a 403 HTTP response code
```

File: tests/functional/features/nhs_number_allocate.feature [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		And product grants access to proxy
+
+		When I request an NHS number allocation
+
+		Then I get a 403 HTTP response code
+		And INVALID_METHOD is at issue[0].details.coding[0].code in the response body
+		And Cannot create resource with application-restricted access token is at issue[0].details.coding[0].display in the response body
+		And Your app has insufficient permissions to use this method. Please contact support. is at issue[0].diagnostics in the response body
+
+	Scenario: A healthcare worker can allocate an NHS number
```
Comment (davidhamill1-nhs) [inline tests/functional/features/nhs_number_allocate.feature:27]:
```suggestion
	Scenario: A healthcare worker's request makes it to Spine

File: proxies/live/apiproxy/proxies/default.xml [hunk 5]
```diff
@@ -92,7 +92,7 @@
   </RouteRule>
   <RouteRule name="sync-wrap">
     <Condition>
-      request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}") AND request.header.x-sync-wrapped NotEquals "true"
+      (request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}") AND request.header.x-sync-wrapped NotEquals "true") OR (request.verb = "POST" AND (proxy.pathsuffix MatchesPath "/Patient") AND request.header.x-sync-wrapped NotEquals "true")
```
Comment (davidhamill1-nhs) [inline proxies/live/apiproxy/proxies/default.xml:95]:
```suggestion
       request.header.x-sync-wrapped NotEquals "true" AND (request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}")) OR (request.verb = "POST" AND (proxy.pathsuffix MatchesPath "/Patient"))
```

File: proxies/live/apiproxy/proxies/default.xml [hunk 6]
```diff
@@ -92,7 +92,7 @@
   </RouteRule>
   <RouteRule name="sync-wrap">
     <Condition>
-      request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}") AND request.header.x-sync-wrapped NotEquals "true"
+       request.header.x-sync-wrapped NotEquals "true" AND (request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}")) OR (request.verb = "POST" AND (proxy.pathsuffix MatchesPath "/Patient"))
```
Comment (claudiudumitrascu1-nhs) [inline proxies/live/apiproxy/proxies/default.xml:95]:
```suggestion
       request.header.x-sync-wrapped NotEquals "true" AND ((request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/{id}")) OR (request.verb = "POST" AND (proxy.pathsuffix MatchesPath "/Patient")))
```

Comment (claudiudumitrascu1-nhs) [reply proxies/live/apiproxy/proxies/default.xml:95]: You might need to use some extra round brackets, A AND B OR C != A AND (B OR C)
Might be possible to test with a POST /Patient without the x-sync-wrapped header

Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/proxies/default.xml:95]: Good spot. Interesting. I would have thought there would be an infinite loop in the tests if that were true. But makes sense to make it explicit.

**Review by claudiudumitrascu1-nhs — APPROVED** [review]

**Pull request merged**